osgtest/library/voms.py

```diff
@@ -116,6 +117,59 @@ def destroy_voms_conf(vo):
     shutil.rmtree(vodir, ignore_errors=True)
 
 
+def proxy_direct(username=None, password=None,
+                 cert_path=None, key_path=None,
+                 fqan=f'/{VONAME}/Role=NULL/Capability=NULL'):
+    f"""Generate a user proxy and directly sign VOMS attributes using the test VOMS cert and key
+    - username: owner of the generated proxy file (default osg-test username)
+    - password: proxy password (default: osg-test password)
+    - cert_path: path to the user certificate (default: ~username/.globus/usercert.pem)
+    - key_path: path to the user key (default: ~username/.globus/userkey.pem)
+    - fqan: VOMS attribute (default: /{VONAME}/Role=NULL/Capability=NULL)
+    """
+    if username:
+        user = pwd.getpwnam(username)
+        uid = user.pw_uid
+        gid = user.pw_gid
+    else:
+        username = core.options.username
+        uid = core.state['user.uid']
+        gid = core.state['user.gid']
+
+    if not password:
+        password = core.options.password
+    if not cert_path:
+        cert_path = os.path.join(os.path.expanduser(f'~{username}'), '.globus/usercert.pem')
+    if not key_path:
+        key_path = os.path.join(os.path.expanduser(f'~{username}'), '.globus/userkey.pem')
+
+    # voms-proxy-direct requires that the user cert/key are owned by the user running the command
+    filemap = dict()
+    for src in (cert_path, key_path):
+        dest = tempfile.NamedTemporaryFile()
+        filemap[src] = dest.name
+        shutil.copy2(src, dest.name)
```

Comment:
According to the help, shutil.copy2 copies the permissions from the original file so you will need to chown the files to root after copying.

Reply:
I think we're in the clear:

> Warning Even the higher-level file copying functions (shutil.copy(), shutil.copy2()) cannot copy all file metadata.
>
> On POSIX platforms, this means that file owner and group are lost as well as ACLs. On Mac OS, the resource fork and other metadata are not used. This means that resources will be lost and file type and creator codes will not be correct. On Windows, file owners, ACLs and alternate data streams are not copied.

Reply:
Just tested it and it looks like you're right: even as root, ownership isn't preserved. Seems like an odd thing to omit though I guess this time it's in our favor.